Sovereign-Labs/sovereign-sdk
# Create a more complete demo-nft-module with collections, metadata uri, freezing, updates, supply

## Overview
`demo-nft-module` has been enhanced to include the concept of Collections, NFTs and calls to create, modify, freeze collections and NFTs and to transfer NFTs. Existing `demo-nft-module` has been moved to `simple-nft-module` since it still serves as a very simply and easy to understand tutorial. `simple-nft-module` focuses more on the sovereign-sdk usage aspects, while `demo-nft-module` aims to be a more ready implementation to plug-n-play and customize if necessary

## Changes
- `demo-nft-module` has no genesis
- `README.md` added to explain the basic concepts, calls provided as well as example `sov-cli` commands to interact with the `demo-nft-module` on a running rollup
- `demo-nft-module` has also been added to runtime, so it would be a permanent module

## Further changes
The offchain component for indexing the NFTs into an offchain store will be part of a second PR.

## Test cases added
- Create Collection
- Mint NFT to created collection
- Mint NFT to non-existent collection
- Update a collection
- Freeze a non existent collection
- Freeze Collection
- Update collection uri for frozen collection
  - assert that the collection uri hasn't been changed
  - assert that supply hasn't been modified
- Supply change after minting an NFT to a Collection
- Minting to a Frozen Collection
  - ensure supply hasn't changed
- Transferring an NFT from a non-owner address
- Transferring NFT with non-existent token id
- Update NFT token_uri
- Freeze NFT
- Update NFT token uri for frozen NFT
